### PR TITLE
feat (current usage): handle current usage case for pay in advance charge

### PR DESCRIPTION
**Pull request by lovrocolic**
## Context

As part of new charge pay in advance feature, current usage logic is changing for charge pay in advance

## Description

This PR covers new logic where maximum value in current period is returned for current usage if charge is paid in advance

### CODE REVIEW

File: app/services/billable_metrics/aggregations/base_service.rb [hunk 1]
```diff
@@ -43,6 +58,26 @@ def sanitized_name(property)
       def sanitized_field_name
         sanitized_name(billable_metric.field_name)
       end
+
+      def handle_in_advance_current_usage(total_aggregation)
+        if previous_event
+          BigDecimal(previous_event.metadata['max_aggregation']) >
```
Comment (jdenquin) [inline app/services/billable_metrics/aggregations/base_service.rb:64]: Lint/Void: Operator `>` used in void context.

File: app/services/billable_metrics/aggregations/base_service.rb [hunk 2]
```diff
@@ -43,6 +58,26 @@ def sanitized_name(property)
...
+          BigDecimal(previous_event.metadata['current_aggregation'])
```
Comment (jdenquin) [inline app/services/billable_metrics/aggregations/base_service.rb:65]: Layout/MultilineOperationIndentation: Use 2 (not 0) spaces for indenting an expression spanning multiple lines.

File: app/services/billable_metrics/aggregations/base_service.rb [hunk 3]
```diff
@@ -43,6 +58,26 @@ def sanitized_name(property)
...
+
+          aggregation = total_aggregation -
+                        BigDecimal(previous_event.metadata['current_aggregation']) +
+                        BigDecimal(previous_event.metadata['max_aggregation'])
+
+          result.aggregation = aggregation
+          result.current_usage_units = total_aggregation
```
Comment (jdenquin) [inline app/services/billable_metrics/aggregations/base_service.rb:72]: Style/IdenticalConditionalBranches: Move `result.current_usage_units = total_aggregation` out of the conditional.

File: app/services/billable_metrics/aggregations/base_service.rb [hunk 4]
```diff
@@ -43,6 +58,26 @@ def sanitized_name(property)
...
+        else
+          result.aggregation = total_aggregation
+          result.current_usage_units = total_aggregation
```
Comment (jdenquin) [inline app/services/billable_metrics/aggregations/base_service.rb:75]: Style/IdenticalConditionalBranches: Move `result.current_usage_units = total_aggregation` out of the conditional.

File: app/services/invoices/calculate_fees_service.rb [hunk 5]
```diff
@@ -108,11 +108,20 @@ def create_charges_fees(subscription, boundaries)
       subscription
         .plan
         .charges
+        .includes(:billable_metric)
         .joins(:billable_metric)
         .where(invoiceable: true)
         .where
         .not(pay_in_advance: true, billable_metric: { recurring: false })
         .each do |charge|
+
```
Comment (jdenquin) [inline app/services/invoices/calculate_fees_service.rb:117]: Layout/EmptyLinesAroundBlockBody: Extra empty line detected at block body beginning. (https://rubystyle.guide#empty-lines-around-bodies)

File: app/services/invoices/calculate_fees_service.rb [hunk 6]
```diff
@@ -108,11 +108,19 @@ def create_charges_fees(subscription, boundaries)
       subscription
         .plan
         .charges
+        .includes(:billable_metric)
```
Comment (vincent-pochet) [inline app/services/invoices/calculate_fees_service.rb:111]: :+1:

File: app/services/invoices/calculate_fees_service.rb [hunk 7]
```diff
@@ -108,11 +108,19 @@ def create_charges_fees(subscription, boundaries)
...
         .joins(:billable_metric)
         .where(invoiceable: true)
         .where
         .not(pay_in_advance: true, billable_metric: { recurring: false })
         .each do |charge|
+        if charge.pay_in_advance? &&
```
Comment (vincent-pochet) [inline app/services/invoices/calculate_fees_service.rb:117]: What about moving all this conditions in a method, for example `should_create_charge_fee?(charge)`?

Comment (vincent-pochet) [reply app/services/invoices/calculate_fees_service.rb:117]: Maybe also, you could add a small spec for this one. What do you think?

**Review by vincent-pochet — APPROVED** [review]
:rocket:

File: app/services/invoices/calculate_fees_service.rb [hunk 8]
```diff
@@ -108,16 +108,26 @@ def create_charges_fees(subscription, boundaries)
...
+        next if should_not_create_charge_fee?(charge, subscription)
+
         fee_result = Fees::ChargeService.new(invoice:, charge:, subscription:, boundaries:).create
         fee_result.raise_if_error!
       end
     end
 
+    def should_not_create_charge_fee?(charge, subscription)
+      charge.pay_in_advance? &&
+      charge.billable_metric.recurring? &&
```
Comment (jdenquin) [inline app/services/invoices/calculate_fees_service.rb:126]: Layout/MultilineOperationIndentation: Use 2 (not 0) spaces for indenting an expression spanning multiple lines.

File: app/services/invoices/calculate_fees_service.rb [hunk 9]
```diff
@@ -108,16 +108,26 @@ def create_charges_fees(subscription, boundaries)
...
+      subscription.terminated? &&
```
Comment (jdenquin) [inline app/services/invoices/calculate_fees_service.rb:127]: Layout/MultilineOperationIndentation: Use 2 (not 0) spaces for indenting an expression spanning multiple lines.

File: app/services/invoices/calculate_fees_service.rb [hunk 10]
```diff
@@ -108,16 +108,26 @@ def create_charges_fees(subscription, boundaries)
...
+      subscription.upgraded?
```
Comment (jdenquin) [inline app/services/invoices/calculate_fees_service.rb:128]: Layout/MultilineOperationIndentation: Use 2 (not 0) spaces for indenting an expression spanning multiple lines.

File: app/services/invoices/calculate_fees_service.rb (hunk 7 above)
Comment (lovrocolic) [reply app/services/invoices/calculate_fees_service.rb]: added:)

**Pull request merged**